src/java/org/jivesoftware/openfire/session/LocalClientSession.java

```diff
@@ -932,6 +932,26 @@ public void setMessageCarbonsEnabled(boolean enabled) {
      */
     @Override
 	public boolean canProcess(Packet packet) {
+        if (JiveGlobals.getBooleanProperty(DISCARD_CARBON_SPOOF, true)) {
+            Message msg = (Message)packet;
+            if (msg != null) {
+                PacketExtension carbon = msg.getExtension("received", "urn:xmpp:carbons:2");
+                if (carbon != null) {
+                    try {
+                        Element copy = carbon.getElement().element("forwarded").element("message");
+                        JID carbonFrom = new JID(copy.attribute("from").getValue());
+                        if (carbonFrom.asBareJID() != packet.getFrom()) {
```

Comment:
I'm not sure if this is correct. Message stanzas that have a carbon should always be from the intended recipients bare JID, and to a full JID of the intended recipient, correct? (Maybe checking that is enough for spoof prevention?) There are two types of carbons: 'received' carbons, and 'sent' carbons. The 'from' of the forwarded message in a 'received' carbon can be another entity address ("one of your full jids received a message from another user, all of your other full jids receive a carbon).